### PR TITLE
fix(finetune): apply CLI config overrides in finetuning runtime

### LINKED ISSUE #2102: [Bug]: CLI --override flag is ignored by the finetuning command
### Version

0.1.dev

### Which installation method(s) does this occur on?

Source

### Describe the bug.

When running `nat finetune` with the `--override` flag (e.g., `nat finetune --override "finetuning.run_configuration.num_epochs=10"`), the configuration overrides are entirely ignored by the runtime. 

The CLI parser in `nat/cli/commands/finetune.py` correctly captures the flag, but there is a stale comment on line 112: `# TODO: Implement config override logic similar to other commands`. 

As a result, the underlying orchestrator in `finetuning_runtime.py` calls the standard `load_config(run_config.config_file)`, ignoring the captured `run_config.override` tuples entirely.

**Expected behavior:**
The `finetune` command should behave exactly like the `run` or `serve` commands, dynamically injecting CLI overrides into the configuration dictionary prior to Pydantic validation via `load_and_override_config`.

### Minimum reproducible example

```shell
# This command runs, but the override is ignored, and the default epochs from the YAML (or schema) are used instead.
nat finetune --override "finetuning.run_configuration.num_epochs=10"
```

### Relevant log output

<details><summary>Click here to see error details</summary><pre>

[No explicit stack trace; the system silently uses the default config values instead of the provided overrides, failing silently in the background config resolution phase.]

</pre></details>

### Other/Misc.

**Possible Solution**:
I have already audited the codebase and locally validated a fix. The solution is to update `packages/nvidia_nat_core/src/nat/finetuning/finetuning_runtime.py` to use `load_and_override_config` alongside `validate_schema`, properly passing `run_config.override`. I will be opening a PR shortly!

### Code of Conduct

- [x] I agree to follow the NeMo Agent Toolkit Code of Conduct
- [x] I have searched the [open bugs](https://github.com/NVIDIA/NeMo-Agent-Toolkit/issues?q=is%3Aopen+is%3Aissue+label%3Abug) and have found no duplicates for this bug report

### PR DESCRIPTION
**Pull request by pranaysb**
## Description

The `--override` flag in the `nat finetune` CLI command was correctly parsed but previously ignored by the underlying `finetuning_runtime.py` because it used the standard `load_config` method instead of `load_and_override_config`. 

This PR fixes the issue by explicitly merging user-provided overrides into the config dictionary before Pydantic validation, aligning the finetuning runtime's behavior with other NVIDIA NeMo Agent Toolkit commands (like `run` and `serve`). 

- **`packages/nvidia_nat_core/src/nat/finetuning/finetuning_runtime.py`**: Replaced `load_config` with `load_and_override_config` to correctly inject CLI overrides.
- **`packages/nvidia_nat_core/src/nat/cli/commands/finetune.py`**: Removed the stale `# TODO` comment since the logic is now properly handled by the runtime layer.

### Verification
Successfully ran the core test suite locally with the fix applied:
<details><summary>Test Output</summary><pre>
$ uv run --extra test --extra eval pytest packages/nvidia_nat_core
...
=============== 2578 passed, 190 skipped, 37 warnings in 28.32s ================
</pre></details>

Closes #2102 

## By Submitting this PR I confirm:
- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/NeMo-Agent-Toolkit/blob/develop/docs/source/resources/contributing/index.md).
- [x] We require that all contributors "sign-off" on their commits. This certifies that the contribution is your original work, or you have rights to submit it under the same license, or a compatible license.
  - [x] Any contribution which contains commits that are not Signed-Off will not be accepted.
- [x] When the PR is ready for review, new or existing tests cover these changes.
- [x] When the PR is ready for review, the documentation is up to date with these changes.

## Summary by CodeRabbit

* **New Features**
  * The finetune CLI now resolves and applies configuration overrides during startup, producing a fully prepared run configuration.
  * Finetuning can start from either a provided configuration object or a standard config file.
* **Bug Fixes**
  * Configuration is now validated before finetuning begins, with clearer handling of override applicability.
  * Improved final job-status logging for completed, canceled, failed, and pending outcomes.
* **Tests**
  * Added CLI unit tests covering finetuning behavior with and without override options.

### CODE REVIEW

File: packages/nvidia_nat_core/src/nat/cli/commands/finetune.py [hunk 1]
```diff
@@ -109,7 +109,6 @@ def finetune_command(
     # Apply overrides if provided
     if override:
         logger.info("Applying config overrides: %s", override)
```
Comment (ericevans-nv) [inline packages/nvidia_nat_core/src/nat/cli/commands/finetune.py:118]: Move the config override handling from `finetuning_runtime.py` here by calling `load_and_override_config`.

File: packages/nvidia_nat_core/src/nat/finetuning/finetuning_runtime.py [hunk 2]
```diff
@@ -100,10 +100,28 @@ async def finetuning_main(run_config: FinetuneRunConfig) -> None:
...
+        import typing
+        config = typing.cast(Config, run_config.config_file)
+        if run_config.override:
+            config_dict = config.model_dump(by_alias=True)
+            layered_config = LayeredConfig(config_dict)
+            for param_path, value in run_config.override:
+                layered_config.set_override(param_path, value)
+            effective_config = layered_config.get_effective_config()
+            config = validate_schema(effective_config, Config)
+    else:
```
Comment (ericevans-nv) [inline packages/nvidia_nat_core/src/nat/finetuning/finetuning_runtime.py:117]: This duplicates logic that belongs in `finetune.py`. Please move it there so `finetune.py` resolves the config first, then hands the final Config object down to the runtime the same way `start.py` does it.

**Review by ericevans-nv — CHANGES_REQUESTED** [review]
Please add unit tests covering the override being applied correctly.

File: packages/nvidia_nat_core/src/nat/cli/commands/finetune.py (hunk 1 above)
Comment (pranaysb) [reply packages/nvidia_nat_core/src/nat/cli/commands/finetune.py:118]: Done. finetune.py now calls discover_and_register_plugins, load_and_override_config, and validate_schema directly before passing the fully-resolved Config object to the runtime via FinetuneRunConfig.

File: packages/nvidia_nat_core/src/nat/finetuning/finetuning_runtime.py (hunk 2 above)
Comment (pranaysb) [reply packages/nvidia_nat_core/src/nat/finetuning/finetuning_runtime.py:117]: Done. finetuning_runtime.py no longer imports or performs any override logic; it just casts the BaseModel or calls load_config for plain paths.

As an added safety measure beyond the original ask, I also added a guard in the runtime: if FinetuneRunConfig.config_file is ever constructed directly as a raw Path and override is non-empty, it raises a ValueError to fail-loud rather than silently dropping overrides like it did originally. Happy to remove this if you prefer the runtime to stay absolutely minimal, but I figured it was worth closing that correctness gap.

**Comment by pranaysb** [comment]
**@ericevans-nv** Pushed a commit addressing all the review feedback:

- **Config resolution** (plugin discovery, override application, schema validation) now lives entirely in `finetune.py`, mirroring `start.py`.
- **`finetuning_runtime.py`** is back to a thin execution layer — no override/plugin logic.
- **Added a `ValueError` guard** to prevent silent override-dropping in any future direct callers of `finetuning_main`.
- **Added CLI-level unit tests** (via `CliRunner`) covering both the override and no-override paths, asserting the override successfully lands in the `Config` object passed to the runtime.
- **Verified locally** with the full `pytest` suite (2580 passed) plus a manual `nat finetune --override ...` run confirming the override is cleanly parsed and reaches the workflow builder.

Ready for another look whenever you have time.

**Review by ericevans-nv — APPROVED** [review]

**Comment by willkill07** [comment]
@pranaysb please ensure all code is formatted using the `yapf` formatter.

We recommend installing and configuring `pre-commit` so that this isn't a persistent issue.

**Comment by pranaysb** [comment]
> @pranaysb please ensure all code is formatted using the `yapf` formatter.
> 
> We recommend installing and configuring `pre-commit` so that this isn't a persistent issue.

Thank you for the review, its done ! i've run yapf to format the code and pushed the updated commit. I also installed the pre-commit hooks locally so this wont happen again moving forward . let me know if anything else is needed!

**Comment by pranaysb** [comment]
> /ok-to-test [a2f8b55](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/2103/commits/a2f8b55f1328fa9cd16a41dbc66ae96ec83e4419)

@ericevans-nv I checked the failing CI logs and noticed the job failed specifically during the Generate Models List phase with this error: ERROR: packages/nvidia_nat_core/tests/nat/cli/commands/test_finetune.py:0 Issue: Apache copyright header missing

it turns out my new test file only had the short 2-line SPDX header. I checked your repository's custom copyright checker (ci/scripts/copyright.py) and saw that on line 146, it explicitly requires the string Licensed under the Apache License, Version 2.0 (the "License"); to pass.

to make sure it's 100% compliant, I grabbed the exact A2_LIC_HASH 14-line template defined on lines 374-388 of your copyright.py script and added it to the top of test_finetune.py. I've pushed the fix and verified it passes locally!

Could you please give it a quick check and trigger /ok-to-test again when you have a moment?

**Comment by ericevans-nv** [comment]
@pranaysb, please make sure the branch is up to date so we can merge. CI is good to go.

**Comment by willkill07** [comment]
/ok to test [26deb36](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/2103/commits/26deb3603cc4af0b7fd6232b5e53009ca45db617)

**Comment by pranaysb** [comment]
> /ok to test [26deb36](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/2103/commits/26deb3603cc4af0b7fd6232b5e53009ca45db617)

Sorry about that! I accidentally force-pushed a duplicate merge commit right as you triggered the tests, which orphaned the old commit hash. Could you please run  the tests  one more time?
